source/sliding_puzzle.py

```diff
@@ -124,7 +139,88 @@ def _join(self, row: List[PuzzlePiece]) -> List[str]:
     def draw(self) -> str:
         """Draw the puzzle in its current state."""
         output = []
-        for row in self.pieces:
+        for row in self.rows:
             output.extend(self._join(row))
 
         return '\n'.join(output)
+
+    def shuffle(self) -> None:
+        """Shuffle the puzzle by randomizing the order of the pieces"""
+        for row in self.rows:
+            random.shuffle(row)
+        puzzle_heigth, puzzle_width = len(self.rows), len(self.rows[0])
+        # empty one of the rows
+        x = random.choice(range(puzzle_heigth))
+        y = random.choice(range(puzzle_width))
+        self.rows[x][y].clear()
+
+    def _get_empty_piece_position(self) -> PiecePosition:
+        x, y, index = [
+            (x, y, piece.index)
+            for x, row in enumerate(self.rows)
+            for y, piece in enumerate(row)
+            if piece._data == []
+        ][0]
+        return PiecePosition(x=x, y=y, index=index)
```

Comment:
Confusing enough, the puzzle is arranged in a list of rows (y).. each row being a list of columns (x).
```
[
  ['a', 'b'],
  ['c', 'd'],
]
```
`'c'` has the indexes: `[1][0]` the 1 represents the second row `↓`, the 0 represents the first column `→`.

What I am saying is that these variables should be flipped

Reply:
columns(y) and rows(x), the y axis is horizontal

Reply:
Yes that is what I mean, the Y axis should be vertical.
```
Y
↑
|
+-→ X
```

Reply:
I meant to write "vertical" not horizontal, columns are vertical